### PR TITLE
HTTP fetch abort_recipe_on_fail attribute

### PR DESCRIPTION
**Pull request by janjurca**
Hello 
I've added feature to http fetch tag. Now attribute abort_recipe_on_fail="on/off" can be placed in there. 
If fetch of this task fail then the taks and all subsequent tasks in recipe are aborted.

### CODE REVIEW

**Comment by jbastian** [comment]
LGTM

**Comment by janjurca** [comment]
@jbastian  
> LGTM

Thanks for review. I've removed the draft status from the pull request. Can you aprove it? Thanks!

File: src/task.c [hunk 1]
```diff
@@ -116,6 +116,7 @@ fetch_finish_callback (GError *error, guint32 match_cnt,
         } else {
             g_propagate_error (&task->error, error);
             task->state = TASK_COMPLETE;
+            task->fetch_succeeded = false;
```
Comment (cbouchar) [inline src/task.c:119]: What if you did,
          app_data->aborted = ABORTED_TASK;
          g_cancellable_cancel(app_data->cancellable);
THEN  (see comments further)

File: src/task.c [hunk 2]
```diff
@@ -1281,6 +1294,10 @@ task_handler (gpointer user_data)
       break;
     }
     case TASK_NEXT:
+      if (!task->fetch_succeeded && task->abort_recipe_on_fail) {
```
Comment (cbouchar) [inline src/task.c:1297]: With this change, it will proceed to get the next task but I believe your intent is to quit the current task after failing to perform the fetch operation for it.  this does not make sense.

**Comment by cbouchar** [comment]
I'm concerned about your changes.  You're injected another path for error handling which may circumvent existing path.  FYI the documentation https://restraint.readthedocs.io/en/latest/results.html will be affected by your changes.

File: src/task.c [hunk 3]
```diff
@@ -1124,9 +1127,19 @@ task_handler (gpointer user_data)
               // If neither started nor finished then fetch the task
               restraint_task_status (task, app_data, "Running", NULL, NULL);
               result = G_SOURCE_REMOVE;
-              g_string_printf(message, "** Fetching task: %s [%s]\n", task->task_id, task->path);
-              app_data->fetch_retries = 0;
-              task->state = TASK_FETCH;
+              if(app_data->state == RECIPE_ABORT){
```
Comment (cbouchar) [inline src/task.c:1130]: I don't think you need these changes.

File: src/task.h [hunk 4]
```diff
@@ -87,6 +87,9 @@ typedef struct RstrntTask {
     /* Whether to keep task changes */
     gboolean keepchanges;
     gboolean ssl_verify;
+    gboolean abort_recipe_on_fail;
+    gboolean fetch_succeeded;
```
Comment (cbouchar) [inline src/task.h:91]: With what I propose, you shouldn't need this fetch_succeeded var.

File: src/recipe.h [hunk 5]
```diff
@@ -37,6 +37,7 @@ typedef enum {
     RECIPE_RUNNING,
     RECIPE_FAIL,
     RECIPE_COMPLETE,
+    RECIPE_ABORT,
```
Comment (cbouchar) [inline src/recipe.h:40]: you shouldn't need this

**Comment by cbouchar** [comment]
Proposal 2:  I don't have your full picture of the problem you are trying to solve but I suspect this problem is with
one task that's giving you an issue with fetch retry.  Instead of changing Restraint, couldn't you write a task following your task which checks whether your fetch operation is on disk.  If not, you could run rstrnt-abort().  This seems like the most least invasive solution.

File: src/task.c (hunk 1 above)
Comment (janjurca) [reply src/task.c]: Hi @cbouchar And thanks a lot for comment.

Your are right this solution would be better and in fact it was the first solution I have tried. But it turned out not to be working. In this solution the task is corectly aborted with the correct abort message. 
I've reimplemented this solution in this commit (https://github.com/restraint-harness/restraint/pull/295/commits/325833750a4a2b51ebd789b4ca8cfb6069535a73) and tried to run it beaker. And even though the task is aborted correctly the whole recipe just continue to run and the following tasks are executed as normal. 
I haven't been able to solve this issue in the past so I've come up with the solution you've been reviewing.

So the actual problem with this solution is that the recipe is not aborted and conitues to run even though the RESTRAINT_ERROR is properly set. 
I'll look at this aborting issue once again and try to solve it, but If you have any idea why the recipe is not aborted I'd like to hear your insights.

Thanks!

File: src/task.c (hunk 2 above)
Comment (janjurca) [reply src/task.c]: Yes I'd like to end on the aborted task. But I thought it should be handled by this:
            g_set_error(&task->error, RESTRAINT_ERROR, RESTRAINT_TASK_RUNNER_ABORTED, "Aborted due to fetch failure");
But it just aborts curent task not the whole recipe. So this was the mechanism to abort all the following tasks manualy with coresponding log message.

Is there a mechanizm that could abort whole recipe? I tried to skip all the following tasks in a list in the recipe. But it did not work with Beaker properly. Beaker was stuck wating for results from those skiped tasks.

File: src/task.c (hunk 3 above)
Comment (janjurca) [reply src/task.c]: Yes if the whole recipe abort would work than it wouldnt be necesary to have this.

File: src/task.c (hunk 1 above)
Comment (cbouchar) [reply src/task.c]: Not sure you saw my 2nd proposal.  This sounds like an extreme corner case for a specific area in your recipe.  It seems more reasonable to create a task to check if your requested test got installed. if not just call rstrnt-abort? Messing with restraint will affect all users and you'd be surprised the bugs that can surface.

Comment (janjurca) [reply src/task.c]: Yes, I saw your second proposal. If there won't be any possibility to push this update upstream, we would need to implement some kind of wrapper. However, even this wrapper could fail to fetch data, which would result in the same outcome as if there were no wrapper at all.

Please take a quick look at my latest changes. I've refactored the code to minimize disruption and make the new feature as optional as possible for the end user.

Regarding this feature, you may see it as an extreme corner case. But is it? All we want to do is abort the recipe if there is a fetch error in some task. As it stands, the user currently has no way to do that in restraint. This provides the user with an option—not a necessity—to abort the recipe, which could result in saving computation time and stopping tasks that would produce irrelevant data.

And yes, I know there's always the possibility that some bugs could surface, but as I've said, this feature is designed to be as optional as possible.

Thanks a lot for your time!

Comment (cbouchar) [reply src/task.c]: My suggestion wasn't to retry a fetch but to check if the directory is present.  if not, abort.  I'm concerned about disrupting other engineers.  BTW I'll be on PTO next week.

Comment (janjurca) [reply src/task.c]: After discussion, let postpone this pull request for Q1. Thanks!
